### ISSUE TITLE
$resource: properties set in transformResponse of actions with isArray===true are swallowed

### ISSUE THREAD

**Issue by chris-33** [angular/angular.js#14797]
***Note*: for support questions, please use one of these channels: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#question. This repository's issues are reserved for feature requests and bug reports.**

**Do you want to request a *feature* or report a *bug*?**
Bug.

**What is the current behavior?**
The bug is in `$resource`. It affects the use of `transformResponse` with actions that have `isArray === true`. If these set properties on the `data`, they are not reflected in the final result.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (template: http://plnkr.co/edit/tpl:yBpEi4).**

See plunk here: http://plnkr.co/edit/Yj3p4srwcToTrAgom3KT?p=preview

Basically, create a resource with a custom `query` action:
```
var Resource = $resource(
    '/foo/:id', 
    { id: '@id' }, 
    { query: {
        method: 'GET',
        isArray: true,
        transformResponse: [ angular.fromJson, function(data) {
            data.foo = 'bar';
            return data;
        }
    }
}
```
Results returned by `Resource.query` will not have the `foo` property. 

**What is the expected behavior?**
Extra properties set on the data should be carried over to the result returned by a resource's action methods.

**What is the motivation / use case for changing the behavior?**
Not in line with expected behavior for `transformResponse`. My particular use case is I need to set meta information on the array that is carried in the headers. I suspect this bug also makes it impossible to wrap arrays in proxies.

**Which versions of Angular, and which browser / OS are affected by this issue? Did this work in previous versions of Angular? Please also test with the latest stable and snapshot (https://code.angularjs.org/snapshot/) versions.**
1.5.0, 1.5.6, Firefox under Ubuntu 16.04 (don't think it makes a difference, though)

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**
In [`resource.js` at line 788](https://github.com/angular/angular.js/blob/master/src/ngResource/resource.js#L788), add code that copies over properties from `data` to `value` while not messing with all the array entries that have just been pushed to `value`.

**Comment by Narretz** [comment]
Resource has a very specific way of handling things so I don't think it can be expected that it accounts for setting metadata on arrays. I'm also not convinced that it is common to set metadata propeties directly on the array. I'll tag this as a feature but I'm not in favor of adding this.

**Comment by gkalpak** [comment]
I agree that this is not a strong enough usecase for changing how `transformResponse` works.
Besides, `tranformResponse` is for transforming the received data - nothing more.

That said, it is already possible (albeit under-documented) to do what you want, by using a response `interceptor`. The (undocumented) trick is that the instance (e.g. the returned array in this case) is available on the `response` parameter passed response interceptors as `response.resource`.

[Updated plnkr][1]

--
Closing, since `transformResponse` is working as expected and what you want is already possible (by other means). FWIW, we are aware that the whole transforms/interceptors story for `$resource` is less than ideal and can be confusing/unintuitive and we plan to fix it for 1.6.

[1]: http://plnkr.co/edit/DsoL2jDVOSl6iSRpUkgI?p=preview

**Comment by chris-33** [comment]
@gkalpak Yeah, that works. :) Thank you for the workaround!
